### PR TITLE
Feat rotation of the state controller

### PR DESCRIPTION
**Pull request by lunfardo314**
state controller rotation (aka committee) function implemented in the core

### CODE REVIEW

File: client/chainclient/chainrecord.go [hunk 1]
```diff
@@ -1,6 +1,8 @@
 package chainclient
 
-import "github.com/iotaledger/wasp/packages/registry/chainrecord"
+import (
+	"github.com/iotaledger/wasp/packages/registry_pkg/chainrecord"
```
Comment (kape1395) [inline client/chainclient/chainrecord.go]: Why do we rename `packages/registry` to `packages/registry_pkg`?
The core functionality is in this package, compared with the corresponding plugin.
Maybe it is better to rename the `plugins/registry` to `plugins/registry_plugin` if that the reason?

File: packages/coretypes/coretypes_test.go [hunk 2]
```diff
@@ -6,26 +6,28 @@ package coretypes
...
+
 	"github.com/iotaledger/goshimmer/packages/ledgerstate"
 	"github.com/iotaledger/hive.go/crypto/ed25519"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
 func TestChainID(t *testing.T) {
-	chid := RandomChainID()
+	chid := chainid.RandomChainID()
```
Comment (kape1395) [inline packages/coretypes/coretypes_test.go]: Maye these tests should be moved to the chainid package?

File: packages/coretypes/registry.go [hunk 3]
```diff
@@ -31,3 +40,26 @@ type PeerNetworkConfigProvider interface {
 	Neighbors() []string
 	String() string
 }
+
```
Comment (kape1395) [inline packages/coretypes/registry.go:43]: I'm not sure, if it is OK to move these interfaces to coretypes. Coretypes got a lot of dependencies in this way.

File: packages/coretypes/rotate/rotate.go [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	}
+	essence, _, err := txb.BuildEssence()
+	if err != nil {
+		return nil, err
+	}
+	return essence, nil
+}
+
+//
+//func MakeStartRotateRequestTransaction(chainID *chainid.ChainID, newStateAddress ledgerstate.Address, inputs []ledgerstate.Output, senderKeyPair *ed25519.KeyPair) (*ledgerstate.Transaction, error) {
```
Comment (kape1395) [inline packages/coretypes/rotate/rotate.go]: Is that still needed here?

File: packages/testutil/testpeers/testkeys.go [hunk 5]
```diff
@@ -1,6 +1,3 @@
-// Copyright 2020 IOTA Stiftung
```
Comment (kape1395) [inline packages/testutil/testpeers/testkeys.go]: Why the header was removed?

File: packages/testutil/testkey/genkey.go [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package testkey
```
Comment (kape1395) [inline packages/testutil/testkey/genkey.go:1]: Why this function is in its own separate file?

File: packages/vm/core/governance/impl.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+	newStateControllerAddr := par.MustGetAddress(ParamStateControllerAddress)
+	// check is address is allowed
+	amap := collections.NewMapReadOnly(ctx.State(), StateVarAllowedStateControllerAddresses)
+	a.Require(amap.MustHasAt(newStateControllerAddr.Bytes()), "rotateStateController: address is not allowed as next state address: %s", newStateControllerAddr.Base58())
+
+	if !newStateControllerAddr.Equals(ctx.StateAnchor().StateAddress()) {
+		// rotate request to another address has been issued. State update will be taken over by VM and will have no effect
+		// By setting StateVarRotateToAddress we signal the VM this special situation
+		// StateVarRotateToAddress value should never persist in the state
+		ctx.State().Set(StateVarRotateToAddress, newStateControllerAddr.Bytes())
```
Comment (kape1395) [inline packages/vm/core/governance/impl.go:36]: This looks like breaking abstractions.
Set a variable that should be not persisted.

File: packages/vm/core/governance/impl.go [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+	// - either there's no need to rotate
+	// - or it just has been rotated. In case of the second situation we emit a 'rotate' event
+	addrs, err := ctx.Call(coreutil.CoreContractBlocklogHname, coretypes.Hn(blocklog.FuncControlAddresses), nil, nil)
+	a.RequireNoError(err)
+	par = kvdecoder.New(addrs, ctx.Log())
+	storedStateController := par.MustGetAddress(blocklog.ParamStateControllerAddress)
+	if !storedStateController.Equals(newStateControllerAddr) {
+		// state controller address recorded in the blocklog is different from the new one
+		// It means rotation happened
+		ctx.Event(fmt.Sprintf("rotate %s %s", newStateControllerAddr.Base58(), storedStateController.Base58()))
```
Comment (kape1395) [inline packages/vm/core/governance/impl.go:50]: Why is the event needed?

File: packages/chain/chainimpl/chainimpl.go [hunk 9]
```diff
@@ -213,30 +255,41 @@ func (c *chainObj) processStateMessage(msg *chain.StateMsg) {
...
+			// close current committee and consensus
+			cmt.Close()
+			c.consensus.Close()
+			c.setCommittee(nil)
+			c.consensus = nil
+			createCommittee = true
+		}
+	} else {
+		// committee does not exist yet. Must be created
+		createCommittee = true
```
Comment (jorgemmsilva) [inline packages/chain/chainimpl/chainimpl.go]: this could be a separate function, and in the next line we can use:
```go
if c.shoudCreateCommittee(msg) {
...
```

File: packages/chain/chainimpl/chainimpl.go [hunk 10]
```diff
@@ -245,45 +298,26 @@ func (c *chainObj) processStateMessage(msg *chain.StateMsg) {
 		c.log,
 	)
 	if err != nil {
-		c.committee = nil
-		c.log.Errorf("failed to create committee object for state address %s: %v", msg.ChainOutput.GetStateAddress().Base58(), err)
-		return
+		c.setCommittee(nil)
+		return xerrors.Errorf("failed to create committee object for state address %s: %w", addr.Base58(), err)
```
Comment (jorgemmsilva) [inline packages/chain/chainimpl/chainimpl.go:299]: should we retry the committee creation in this case?

File: packages/vm/core/governance/impl.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+	return nil, nil
+}
+
+func addAllowedStateControllerAddress(ctx coretypes.Sandbox) (dict.Dict, error) {
+	a := assert.NewAssert(ctx.Log())
+	a.RequireChainOwner(ctx, "addAllowedStateControllerAddress")
+	par := kvdecoder.New(ctx.Params(), ctx.Log())
+	addr := par.MustGetAddress(ParamStateControllerAddress)
+	amap := collections.NewMap(ctx.State(), StateVarAllowedStateControllerAddresses)
+	amap.MustSetAt(addr.Bytes(), []byte{0xFF})
```
Comment (kape1395) [inline packages/vm/core/governance/impl.go:63]: What is 0xFF?

File: packages/coretypes/request/requestimpl.go [hunk 12]
```diff
@@ -151,7 +154,7 @@ type RequestOnLedger struct {
 	outputObj       *ledgerstate.ExtendedLockedOutput
 	requestMetadata *RequestMetadata
 	senderAddress   ledgerstate.Address
-	solidArgs       dict.Dict
+	params          atomic.Value // this part is mutable
```
Comment (jorgemmsilva) [inline packages/coretypes/request/requestimpl.go:156]: why are the request params mutable?

File: packages/registry_pkg/blobcache_test.go [hunk 13]
```diff
@@ -1,4 +1,4 @@
-package registry
+package registry_pkg
```
Comment (jorgemmsilva) [inline packages/registry_pkg/blobcache_test.go]: I renamed `registry_pkg` to `registry` following this comment by @dessaya : https://github.com/iotaledger/wasp/pull/177#discussion_r642696550

any reason we are reverting this 😅 ?

File: client/chainclient/chainrecord.go (hunk 1 above)
Comment (dessaya) [reply client/chainclient/chainrecord.go]: IMO they should both be just `registry`. There is no name clash since there should be no need to import both at the same time ever.

File: client/chain_record.go [hunk 14]
```diff
@@ -3,8 +3,10 @@ package client
 import (
 	"net/http"
 
-	"github.com/iotaledger/wasp/packages/coretypes"
-	"github.com/iotaledger/wasp/packages/registry/chainrecord"
+	"github.com/iotaledger/wasp/packages/registry_pkg/chainrecord"
+
+	"github.com/iotaledger/wasp/packages/coretypes/chainid"
+
```
Comment (dessaya) [inline client/chain_record.go]: Why the extra blank lines? :cry:

File: docs/docs/coretypes.md [hunk 15]
```diff
@@ -20,7 +20,7 @@ transient because chains can be moved from address to address. The chain color
 is an ultimate identifier of the chain for its lifetime.
 
 Each chain is identified on the ISCP by its _chain ID_, represented by the
-[`coretypes.ChainID`](https://github.com/iotaledger/wasp/blob/master/packages/coretypes/chainid.go)
+[`coretypes.ChainID`](https://github.com/iotaledger/wasp/blob/master/packages/coretypes/chainid/chainid.go)
```
Comment (dessaya) [inline docs/docs/coretypes.md]:
```suggestion
[`chainid.ChainID`](https://github.com/iotaledger/wasp/blob/master/packages/coretypes/chainid/chainid.go)
```

File: packages/chain/chainimpl/chainimpl.go [hunk 16]
```diff
@@ -205,6 +204,49 @@ func (c *chainObj) processPeerMessage(msg *peering.PeerMessage) {
 	}
 }
 
+// processStateMessage processes the unique chain output which exists on the chain's address
```
Comment (dessaya) [inline packages/chain/chainimpl/chainimpl.go]:
```suggestion
// processChainTransition processes the unique chain output which exists on the chain's address
```

File: packages/chain/peermsg.go [hunk 17]
```diff
@@ -5,161 +5,88 @@ package chain
...
-	}
-	if n != ledgerstate.OutputIDLength {
-		return xerrors.Errorf("failed parse OutputID: wrong data length")
-	}
-	return nil
+// Message types for the committee communications.
+const (
+	MsgGetBlock     = 1 + peering.FirstUserMsgCode
+	MsgBlock        = 2 + peering.FirstUserMsgCode
+	MsgSignedResult = 3 + peering.FirstUserMsgCode
```
Comment (dessaya) [inline packages/chain/peermsg.go:24]:
```suggestion
	MsgGetBlock     = 1 + peering.FirstUserMsgCode + iota
	MsgBlock
	MsgSignedResult
```

File: packages/chain/peermsg.go [hunk 18]
```diff
@@ -5,161 +5,88 @@ package chain
...
-		}
-	}
-	if err := msg.FeeDestination.Read(r); err != nil {
-		return err
-	}
-	return nil
+// VMResultMsg Consensus -> Consensus. VM sends result of async task started by Consensus to itself
+type VMResultMsg struct {
+	Task    *vm.VMTask
+	NoError bool
```
Comment (dessaya) [inline packages/chain/peermsg.go]: Any reason to prefer `NoError` instead of just `Error`?
IMO `Error` is slightly better since it does not end up in double negatives (`if !NoError {...}`) and also the zero value (`false`) is a good default.

File: packages/coretypes/assert/util.go [hunk 19]
```diff
@@ -26,6 +28,16 @@ func (a Assert) Require(cond bool, format string, args ...interface{}) {
 	a.log.Panicf(format, args...)
 }
 
-func (a Assert) RequireNoError(err error) {
-	a.Require(err == nil, fmt.Sprintf("%v", err))
+func (a Assert) RequireNoError(err error, str ...string) {
+	a.Require(err == nil, fmt.Sprintf("%s %v", strings.Join(str, " "), err))
+}
+
+func (a Assert) RequireChainOwner(ctx coretypes.Sandbox, name ...string) {
```
Comment (dessaya) [inline packages/coretypes/assert/util.go:40]: This `name` parameter maybe should be a member of the `Assert` object, so all `Require*` methods can benefit from it.

```go
	a := assert.NewAssert(ctx.Log(), "this is the name of the assert")
	a.RequireChainOwner(ctx)
```

File: packages/coretypes/chainid/chainid.go [hunk 20]
```diff
@@ -1,7 +1,7 @@
 // Copyright 2020 IOTA Stiftung
 // SPDX-License-Identifier: Apache-2.0
 
-package coretypes
+package chainid
```
Comment (dessaya) [inline packages/coretypes/chainid/chainid.go:4]: Was this change really necessary?

If `ChainID` really needs to be on its own package, following idiomatic rules we should try to avoid the stuttering (`var chainID chainid.ChainID`). For example calling the package `chain` and the object `ID`: `var chainID chain.ID`.

File: client/chain_record.go (hunk 14 above)
Comment (lunfardo314) [reply client/chain_record.go]: no idea.

File: client/chainclient/chainrecord.go (hunk 1 above)
Comment (lunfardo314) [reply client/chainclient/chainrecord.go]: just followed the trend :)

**Review by kape1395 — APPROVED** [review]
